### DISCUSSION TITLE
Can’t access Documents and ICloud on IPad using Blink Code FS extension

### DISCUSSION THREAD

**Discussion by carstenf** [blinksh/blink#1828] [Support] [answered]
I downloaded blink.sh to my IPad
Typed code .
VSCode opens and I installed Blink-FS
Now I can see Documents and ICloud, but below are now files/folders visible.
What do I need to change to access the files and folders below Documents and ICloud?

second issue, I can’t close code.
I tried cmd+w, cmd+shift+w, shift+cmd+w, no reaction.

im using iOS 16.6

**Comment by yury** [comment, marked as answer]
Hello @carstenf.

Can you email me (yury at blink.sh) ~/.blink/blinkCode.log ?

```sh
cat ~/.blink/blinkCode.log | pbcopy
```
and paste in email.

**Comment by carstenf** [comment]
I send it to you, got it?

**Comment by carloscabanero** [comment]
@carstenf we got the logs, thanks! Would you mind to attach here a screenshot of the structure as shown in VS Code at the moment? Thanks

**Comment by carstenf** [comment]
<img width="834" alt="image" src="https://github.com/blinksh/blink/assets/58643140/581ba04a-2e4d-4e66-b04d-0147563a9851">

**Comment by carstenf** [comment]
That screenshot is ok?

I tried as well to connect to a AWS EC2.
It was connected but I don’t even see a tree, the blue folder icon don’t respond at all.
Super strange.
…and as I said, I can’t quit the code windows, which should be shift+command+W.
I checked the settings and it is that command.
I tried to close a regular terminal and it working.
So it looks like everything with code . is not performing as expected.

**Comment by yury** [comment]
With Blink v16.2.0 we introduced new quick actions bar (double cmd) it has `close` tab action. We are working on improving VSCode shortcuts integration.

**Reply by carstenf** [reply]
Im using only a software keyboard.
I read this in your manual:
„Tapping twice in the iOS Home Bar, or pressing the CMD ⌘ key twice, in Software or Hardware keyboards, will open a contextual menu where you can create or close tabs, open Snips, or configure display modes.'“

I guess it’s you suggestion. 
The tapping twice in the OS home bar will not work, as a single tap already bring up the software keyboard.
The twice tapping of cmd key only toggle between the F1 to F6 and some other additional keys.
It does not bring up the quick action bar.

**Reply by carloscabanero** [reply]
Please make sure to update to the latest Blink version (we just released today). The new Context Bar is new there and there is a new gesture for it (the double tap on Cmd or Home bar).

**Reply by carstenf** [reply]
Jus downloaded it, it’s working now.

**Comment by carloscabanero** [comment]
Hey @carstenf, so I have been taking a look at the logs and your screenshot and I could not see anything wrong, but maybe it is unclear how the extension works.

On local, I see that the extension is actually already serving files, so it looks like it works as expected. Documents and iCloud directory are referring to your Blink Documents and Blink iCloud folder. If you do a `ls iCloud` or `ls Documents`, are they also empty or do they show any files? I think maybe you expect to show your full iCloud folder, but Blink does not have direct access to that. You can use the `link-files` command to link other file structures from other apps.

About the remote over AWS EC2, can you try to connect first using the Files.app? Make sure you specify a Key or Password as automatic login is necessary. And make sure the remote also has sftp enabled.

I'm moving this as a Discussion as it looks to me like this is more a Support request.

**Comment by carstenf** [comment]
Hi Carlos,

to you question:
The terminal in the VSCode window inside blink is not working, I can’t do ls command.
„Terminals are not available in the web editor“

we’re do I find in the files.app the Blink Documents and Blink iCloud folder?
For some other apps I find inside the ICloud folder the app, there is no Blink app there.
The only blink symbol in the whole files.app is the on who will be connected to the server.

Next:
I configured a host for the AWS EC1 and specify a folder in the config host settings. The connection with the Files.app is working. If I tap the blink symbol, I can see the folder I created in the Ubuntu folder on the AWS.

Next I did „ssh hostname“ I’m in the server, everything fine.

Then I started the vscode on the AWS EC1 with the command „code-server“, coder server is running.
Open a new terminal in Blink and use the command „code hostname“, the vscode opens, but I can’t asses any folder, file now. I can‘t even open a terminal inside VSCode, please see the screen shot.
What do I do wrong? I read a dozent times the manual on your homepage and watched different YouTube videos about blink and VSCode. The version I installed was from coder.com.
![image](https://github.com/blinksh/blink/assets/58643140/4d03169e-0d52-4470-a1d6-32516a0a58a9)

**Reply by carloscabanero** [reply]
> Hi Carlos,
> 
> to you question: The terminal in the VSCode window inside blink is not working, I can’t do ls command. „Terminals are not available in the web editor“

> we’re do I find in the files.app the Blink Documents and Blink iCloud folder? For some other apps I find inside the ICloud folder the app, there is no Blink app there. The only blink symbol in the whole files.app is the on who will be connected to the server.

I meant to do the `ls Documents` or `ls iCloud` inside the Blink shell, not the VS Code shell.
iCloud and Documents folder refer to your device. The iCloud and Documents folders will only show if you have any files inside. You can do create files from VS Code or do `touch Documents/test` in the Blink Shell in order to show them.

> Next: I configured a host for the AWS EC1 and specify a folder in the config host settings. The connection with the Files.app is working. If I tap the blink symbol, I can see the folder I created in the Ubuntu folder on the AWS.
> 
> Next I did „ssh hostname“ I’m in the server, everything fine.
> 
> Then I started the vscode on the AWS EC1 with the command „code-server“, coder server is running. Open a new terminal in Blink and use the command „code hostname“, the vscode opens, but I can’t asses any folder, file now. I can‘t even open a terminal inside VSCode, please see the screen shot. What do I do wrong? I read a dozent times the manual on your homepage and watched different YouTube videos about blink and VSCode. The version I installed was from coder.com. ![image](https://user-images.githubusercontent.com/58643140/261448292-4d03169e-0d52-4470-a1d6-32516a0a58a9.png)

So let's start with one note to understand how Blink Code works. Blink Code is just a layer on top of the instance of VS Code that you decide to connect to. When you do `code .`, or `code host:~/`, Blink Code will use vscode.dev and the Blink-FS file extension to connect to the local or remote filesystem. When you do `code https://route_to_remote_instance`, Blink Code will just connect to that remote Blink Code instance, no Blink-FS required. Note that Blink Code uses the `https` or `host:` to decide what protocol to use for the connection.

If you want to connect to your remote using SFTP, you do not need to run code-server in the remote, but then the terminal won't be available (as the connection will happen through vscode.dev). `code hostname:~/`

If you run code-server, you want to connect to that code server running instance, so no SFTP. So you need to pass it as a URL `code https://url_for_code_server`. My preference is to install the regular MS VS Code instead of code-server. They now include `code tunnel` option, that will give you a URL that is easy to connect to.

**Comment by carstenf** [comment]
Hi Carlos, 

I just want to use my iCloud or documents environment like on the MacBook, and code on the IPad.
Basically switching between MacBook and IPad. Now I’m not sure if this is possible with Blink.

Maybe you can add a small explanation in the code part of you online manual, how to install VSCode on a external AWS or other server. Than how to connect it with blink, step by step. Just you perfered version of which version of VSCode to use and which kind of connection would be the best. In this case everyone could follow it and there would be less confusion.
Tomorrow I will try your suggestions.
Thankx

**Reply by carloscabanero** [reply]
Been working on that this week, I will push it tomorrow.

**Reply by carstenf** [reply]
Hi Carlos 

I installed VSCode on the AWS EC2.

If I understood you correctly the best way to do is:
- start VSCode on the AWS EC2 with `code tunnel —name VSCode1`
- Start in blink shell `code .`
- Install blink-fs extension inside VSCode window
- Connect inside VSCode window to the ssh remote tunnel to VSCode1

I now can see the Document and iCloud folder inside the VSCode window.
so far correct?

But I do not get a Terminal inside the VSCode window, it gives an error.
(I just got for one secession a working terminal with that error, but now it’s gone again)

![image](https://github.com/blinksh/blink/assets/58643140/ea96c6c7-4aa2-4e89-be53-65f9d78df46a)

**Reply by carloscabanero** [reply]
No, I don't think that's what you want. If you start VS Code on the remote, you don't want to connect to the "local" file system using `code .`. When you run VS Code on the remote, you should receive a URL that you can use to connect from your side code https://whatever_url`. That is what I did here: https://youtube.com/shorts/2QDsyR0kvo0?feature=share

Note that when you are connected to the remote instance, Blink Code is just a facade trying to give you a better experience than your browser. Blink-FS and all of that is not necessary, as you are connected to VS Code using their remote tunnels.

Although connecting to your remote the way you did may work, it looks like there may be other additional issues that they may still be ironing out. Ie, the terminal is trying to run in the local shell, but there is no local terminal on your iPad.

**Comment by carstenf** [comment]
Hi Carlos 

i found now an other issue.

I copied some code + data via files.app to the iCloud of Blink inside the VSCode - worked fine.
The folder was named Test inside the iCloud folder.
Now I run the a python program inside the test folder which requires to read csv data which are as well inside the test folder. It looks like it does not find these data sets. 
I don’t know were that folder is located, private/…/test

please see the screenshot below

![image](https://github.com/blinksh/blink/assets/58643140/8d6fddbc-222c-4d67-aa02-7ff561b3b810)

**Reply by carloscabanero** [reply]
Hi again! So in this scenario again, the code should be in the remote where your Python and all of that is. You connect to that remote VS Code instance, and not to "code .", and you do not need Blink-FS for this.

That being said, this is a case I had not tried myself and there is one very interesting thing which you discovered. Even though the code is local to your iPad, your AWS instance is accessing it through VS Code. But I think that is only for that file, and as you can see it doesn't have access to the rest of the file system, hence the error message you see.

But again, you may be doing things more difficult than they are. Just use the url you get from running code in the remote and pass it to `code https://...`.

**Reply by carstenf** [reply]
Ok, than please tell me how I should do it 😃 ...
I want to use the file.app connection to copy files to the iCloud inside blink.
Then I want to run everything what's in the folder I copied to iCloud blink.
And off course a funktional terminal inside VSCode.

That means I need to use code ., else I loose the files.app advantage.

If I do 'code . Host' it's not working. 
The only possible solution which worked, was to start code . and than connect with a tunnel to the server. And to have a tunnel I need to run code tunnel -name... on the server

Some other suggestions?

**Reply by carloscabanero** [reply]
Ok, so let's see if I do a better job now at explaining this and maybe I can even come up with an article on it.

The iPad is not able to run any code*, so "storing" the code in the iPad or "serving" your code is just inefficient. It is useful if you just want to do quick edits, etc... but that is all there is to it. So you need to forget about the iCloud and stuff like that because those are not steps of anything or a goal, and they are wrong. Your goal is that you have python code that you need to run and work on.
*_This is not fully true, there are apps that specialize on running Python code (like Jupyter notebooks, etc...) and will create a proper environment with its limitations. Blink focuses on Remote execution, giving you tools to connect to Remotes and run code there, and to offer a "window" so that the Remote feels like local._

So you have code that you want to run and edit from your iPad. And access a "terminal" inside VS Code for that "Remote" or machine or whatever. 

You already have created an AWS instance for this task (the Remote). Instead of having your code "local", on the iPad, download the "code" in the Remote, where the Python interpreter is. You can connect to the remote using Blink's ssh or mosh. In the remote, run `code tunnel` and it should give you a URL. Copy the URL and in a new Blink Shell, run `code https://the_url_you_got`. When code opens, you will see it is running _already connected to the remote_ and it has access to the full "remote file system", including all the code and tooling.

Now, Blink also offers a few more advantages to make the Remote feel like local. For example, you can access the Remote through Mosh instead of the VS Code terminal - this has the advantage that you will have a permanent connection to it. Also, Blink can expose the Remote file system through the Files.app, so you can copy the files you placed in the iCloud folder to the Remote by drag and dropping them, or you can use other apps to edit those Remote CSV files (https://docs.blink.sh/advanced/files-app). And this is all to bring the Remote to the local, not the local to the Remote.

Let me know if this is now more clear, if not I'm happy to jump on a call with you or chat on Discord to clarify and guide you step by step.

PS: I know that the misunderstanding comes from you finding out something actually quite cool, that the local code is being served to the Remote, and even being executed. I had no idea about this, but ignore it because it is not fully complete and it looks like a dead end that I will investigate. But for now, put the code in the remote, that is what we recommend.

**Reply by carloscabanero** [reply]
One more thing, you said in a different thread "switching between MacBook and iPad". In that case, treat your "MacBook" as the remote (instead of AWS). You can have your python code there. And then run `code tunnel` in the MacBook, and use the resulting URL to connect Blink to it. No need to have any code in the iPad, in Blink or in iCloud.

**Comment by carstenf** [comment]
“But for now, put the code in the remote, that is what we recommend.”
This is what I was trying in the end.

From my understanding, if I use `code .`, I can use the files.app and put the the code into the remote (see my last screen shot) and as you are writing above. 
(If I use `code .` I understood, I can not use the terminal - ok fine, then I use blink with mosh instead)

**Now, how is the command using `code .` to connect to a remote?**
(As I understood, the remote doesn’t need to run code, or code tunnel.)
`code . Hostename` does not work, tried it.
`code . URL` not possible as I did not run code tunnel on the remote
`code . ` and then how do I connect to the remote if I did not start the code tunnel?  

Sounds a bit silly that question, maybe I overlooked the answer - sorry

**Reply by carloscabanero** [reply]
> From my understanding, if I use `code .`, I can use the files.app and put the the code into the remote (see my last screen shot) and as you are writing above. (If I use `code .` I understood, I can not use the terminal - ok fine, then I use blink with mosh instead)
> 
This is not correct. When you use `code .`, you are telling Blink acode to open vscode.dev and use Blink-FS to connect to the Local file system at the current directory (hence the dot). What you have in that screenshot is Blink with the code in the local file system. The "dot" is the "current directory", it would be like saying `code ./` or `code .ssh` to open the ssh folder.

> **Now, how is the command using `code .` to connect to a remote?** (As I understood, the remote doesn’t need to run code, or code tunnel.) `code . Hostename` does not work, tried it. `code . URL` not possible as I did not run code tunnel on the remote `code . ` and then how do I connect to the remote if I did not start the code tunnel?
> 
> Sounds a bit silly that question, maybe I overlooked the answer - sorry

That is correct, you can open code and connect over SFTP to the remote using Blink-FS. No tunnel required in this case, it goes through vscode.dev, and then you can use mosh to shell into your remote. You just _do not need the dot in this case._
`code <host name>:~/`
`code <host_name>:<remote_path>`
Note that `code <host_name>` alone won't work, because Blink Code will interpret it as a local path. What signals 'remote over sftp' is the semicolon `<host-name>:` 

And if you want to connect to a remote VS Code session, over tunnel, gitpod, etc:
`code https://the_remote_route`

In the latest Blink, we now have [Snips](https://docs.blink.sh/basics/snips). Snips can help you fill out code commands:
<img width="618" alt="image" src="https://github.com/blinksh/blink/assets/536729/01cc4fe3-c7b3-42b1-8c34-e60881eff08d">

All this information is already in the docs: https://docs.blink.sh/advanced/code
But I still owe you an article about installing code in the remote.

**Comment by carstenf** [comment]
Ok, we are getting there…

code <host_name>:<remote_path> opens vscode on the remote we’re i stored my code.
In my case I used the alias for the configured server in `config`, with his is `vscode`
The path in the remote ist `/home/ubuntu/test` so I did
`code vscode:/home/ubuntu/test` which looks fine. Using `~ /home/ubuntu/test ` lands me somewhere… does not matter.
Now the issue is that vscode.dev in blink finds the right language, but does not find the kernel to run the code…
Looks like I can not run the vscode.dev on my iPad and use the installed python compiler on the remote?

see the screen shot 1 with `code vscode:/home/ubuntu/test`

and the second screen shot using code `https://vscode.dev/tunnel/vscode1 ` were everything works fine

`code vscode:/home/ubuntu/test`
![image](https://github.com/blinksh/blink/assets/58643140/dc1aa35d-4a1d-4bd5-9438-6ea09a90e1eb)


`https://vscode.dev/tunnel/vscode1 `
![image](https://github.com/blinksh/blink/assets/58643140/fb3fdc32-9306-4e61-b054-72ef8e11415d)

**Reply by carloscabanero** [reply]
Yeah, so it looks like you are now going full circle for all the ways you can run Blink Code and the limitations for each.

- This is expected. When you run Blink Code and connect over Blink-FS to a remote SFTP file system, only those files are being served. And you are still within the vscode.dev limitations, so you won't have access to the interpreter as you just noticed. There may be some language functionality and auto completion, but very limited. It is still useful in certain cases if you use the terminal as a companion (ie, JS development). That's why at the start of the full thread I just asked you to connect to the remote VS Code instance `code https://whatever-url-it-gave-you`.
- The issue with the Home folder not resolving may be a bug on our side. I will run a few quick tests and ask you for logs if necessary.

So yeah, that is all there is, the full circle:
- code for local files with Blink-FS is useful to edit local files, but you are still within the iPad sandbox.
- Code for remote file editing using Blink-FS is useful to edit remote files, even on limited devices like a Raspberry Pi. But you are still within the vscode.dev limitations.
- Code connecting to a remote is the recommended way for a full VS Code experience as you are used to. No need to use Blink-FS.

PS: You did discover one interesting scenario, connecting to the remote VS Code instance but using Blink-FS with the local file system. It seems like it has its limitations but we need to investigate if we can take it to its full potential.

**Comment by carstenf** [comment]
I went back to coder / coder-sever, as it has a faster connection.
The tunnel from VSCode ist very slow to reconnect, if one changes a folder.  
That’s what I’m using now, which works fine.
`code http://35.242.210.242:8080`
The code-server can be set to start automatically if the VM is started.

Next I installed rclone on the VM to connect and sync (automatically) with a could drive, like google drive, unfortunately iCloud is not supported. 

One item on the wish list would be to press long the space key on the iPad software keyboard to control the curser inside the code window, any chance to get working?
(Using the 4 small arrows on the keyboard is a bit difficult)
